src/main/scala/org/zalando/spearheads/innkeeper/api/JsonProtocols.scala

```diff
@@ -1,122 +1,164 @@
+    override def read(json: JsValue): RegexPathMatcher = ???
   }
 
-  val endpointFormat = jsonFormat(Endpoint.apply, "hostname", "path", "port", "protocol", "type")
+  implicit object StrictPathMatcherFormat extends RootJsonFormat[StrictPathMatcher] {
+    override def write(hm: StrictPathMatcher): JsValue =
+      JsObject(Map("match" -> JsString(hm.matcher),
+        "type" -> JsString(MatcherType.STRICT)))
 
-  implicit object EndpointFormat extends JsonFormat[Endpoint] {
-    override def write(obj: Endpoint): JsValue = endpointFormat.write(obj)
+    override def read(json: JsValue): StrictPathMatcher = ???
+  }
 
-    override def read(json: JsValue): Endpoint = {
-      val endpoint = endpointFormat.read(json)
+  implicit object PathMatcherFormat extends RootJsonFormat[PathMatcher] {
+    def write(pathMatcher: PathMatcher) = pathMatcher match {
+      case RegexPathMatcher(matcher) =>
+        JsObject(Map("match" -> JsString(matcher),
+          "type" -> JsString(MatcherType.REGEX)))
+      case StrictPathMatcher(matcher) =>
+        JsObject(Map("match" -> JsString(matcher),
+          "type" -> JsString(MatcherType.STRICT)))
+    }
 
-      endpoint.copy(
-        port = endpoint.port.orElse(Some(443)),
-        protocol = endpoint.protocol.orElse(Some(Https)),
-        endpointType = endpoint.endpointType.orElse(Some(ReverseProxy))
-      )
+    def read(value: JsValue) = value.asJsObject.getFields("match", "type") match {
+      case Seq(JsString(matcher), JsString(MatcherType.REGEX)) =>
+        new RegexPathMatcher(matcher)
+      case Seq(JsString(matcher), JsString(MatcherType.STRICT)) =>
+        new StrictPathMatcher(matcher)
+      case _ => deserializationError("PathMatcher expected")
     }
   }
 
-  implicit object MatcherTypeFormat extends JsonFormat[MatcherType] {
-    val REGEX = "REGEX"
-    val STRICT = "STRICT"
+  private val matcherFormat = jsonFormat(Matcher, "hostMatcher", "pathMatcher", "methodMatcher", "headerMatchers")
 
-    override def write(matcherType: MatcherType): JsValue = matcherType match {
-      case Regex  => JsString(REGEX)
-      case Strict => JsString(STRICT)
+  implicit object MatcherFormat extends RootJsonFormat[Matcher] {
+
+    private def validate(matcher: Matcher, ex: Exception) = {
+      if ((!matcher.headerMatchers.isDefined || matcher.headerMatchers.get.isEmpty) &&
+        !matcher.hostMatcher.isDefined &&
+        !matcher.methodMatcher.isDefined &&
+        !matcher.pathMatcher.isDefined) {
+        throw ex
+      }
+      matcher
     }
 
-    override def read(json: JsValue): MatcherType = json match {
-      case JsString(REGEX)  => Regex
-      case JsString(STRICT) => Strict
-      case _                => deserializationError("MatcherType expected")
+    override def write(matcher: Matcher): JsValue = {
+      matcherFormat.write {
+        validate(
+          matcher,
+          new SerializationException("At least one of the fields in the matcher should be defined")
```

Comment:
This interface of `validate` creates an exception, that is usually never needed. What about providing a (reference to a) function, that generates (or throws) the exception instead of the exception itself.